lib/membrane_video_compositor_multiple_inputs.ex

```diff
@@ -0,0 +1,294 @@
+    ]
+  )
+
+  def_input_pad(
+    :input,
+    demand_unit: :buffers,
+    availability: :on_request,
+    demand_mode: :auto,
+    caps: {RawVideo, pixel_format: :I420}
+  )
+
+  def_output_pad(
+    :output,
+    demand_unit: :buffers,
+    demand_mode: :auto,
+    caps: {RawVideo, pixel_format: :I420}
+  )
+
+  defmodule Track do
+    @moduledoc false
+
+    @type buffer_t :: Membrane.Buffer.t()
+
+    @type t :: %__MODULE__{
+            buffers: Qex.t(buffer_t),
+            status: :playing | :end_of_stream
+          }
+    defstruct buffers: Qex.new(), status: :playing
+
+    @doc """
+    Checks whether track is empty and can be removed
+    """
+    @spec depleted?(__MODULE__.t()) :: boolean()
+    def depleted?(%__MODULE__{status: status, buffers: buffers}) do
+      status == :end_of_stream and Enum.empty?(buffers)
+    end
+
+    @spec push_frame(__MODULE__.t(), buffer_t) :: __MODULE__.t()
+    def push_frame(%__MODULE__{buffers: buffers} = track, frame) do
+      %__MODULE__{track | buffers: Qex.push(buffers, frame)}
+    end
+
+    @spec pop_frame(__MODULE__.t()) :: __MODULE__.t()
+    def pop_frame(%__MODULE__{buffers: buffers} = track) do
+      %__MODULE__{track | buffers: Qex.pop!(buffers) |> elem(1)}
+    end
+
+    @spec first_frame(__MODULE__.t()) :: buffer_t
+    def first_frame(%__MODULE__{buffers: buffers}) do
+      Qex.first!(buffers)
+    end
+
+    @spec has_frame?(__MODULE__.t()) :: boolean
+    def has_frame?(%__MODULE__{buffers: buffers}) do
+      not Enum.empty?(buffers)
+    end
+  end
+
+  @impl true
+  def handle_init(options) do
+    compositor_module = determine_compositor_module(options.implementation)
+
+    {:ok, internal_state} = compositor_module.init(options.caps)
+
+    state = %{
+      ids_to_tracks: %{},
+      caps: options.caps,
+      compositor_module: compositor_module,
+      internal_state: internal_state,
+      pads_to_ids: {0, %{}}
+    }
+
+    {:ok, state}
+  end
+
+  @impl true
+  def handle_pad_added(pad, _context, state) do
+    state = register_track(state, pad)
+    {:ok, state}
+  end
+
+  defp register_track(state, pad) do
+    {new_id, pads_to_ids} = state.pads_to_ids
+    state = %{state | pads_to_ids: {new_id + 1, Map.put(pads_to_ids, pad, new_id)}}
+
+    %{state | ids_to_tracks: Map.put(state.ids_to_tracks, new_id, %Track{})}
+  end
+
+  @impl true
+  def handle_caps(pad, caps, _context, state) do
+    %{pads_to_ids: {_new_id, pads_to_ids}, internal_state: internal_state} = state
+    id = Map.get(pads_to_ids, pad)
+
+    {:ok, internal_state} =
+      state.compositor_module.add_video(id, caps, %{x: 0, y: 0}, internal_state)
+
+    state = %{state | internal_state: internal_state}
+    {{:ok, caps: {:output, state.caps}}, state}
+  end
+
+  @impl true
+  def handle_process(
+        pad,
+        buffer,
+        _context,
+        %{
+          ids_to_tracks: ids_to_tracks,
+          pads_to_ids: {_new_id, pads_to_ids}
+        } = state
```

Comment:
Don't unpack arguments in the function head unless required to match. Unpack in the function body instead.